### DISCUSSION TITLE
Own theory implementation

### DISCUSSION THREAD

**Discussion by vhavlena** [cvc5/cvc5#9178] [Q&A] [answered]
Dear cvc5 authors,

could I find somewhere a guide on how to implement own decision procedure for a theory in cvc5? In particular, which interface it is necessary to implement, which functions, etc.

Thank you very much,
Vojta

**Comment by martin-cs** [comment]
Hi @vhavlena ,

Thanks for the interest.  Do you want to implement your own solver for an existing theory or do you want to create a new theory?

The first case is somewhat easier, very roughly, you need to replace the relevant child of `Theory`( https://github.com/cvc5/cvc5/blob/main/src/theory/theory.h ) for example https://github.com/cvc5/cvc5/blob/270b81b45afaeaf239b5458bbbe7166ee3ba83dc/src/theory/fp/theory_fp.h#L38 is the floating-point solver.  These make use of a bunch of interfaces in the Theory namespaces but are intended to contain the "decision procedure" part.

If you want to create a new theory entirely, you will need to add the relevant kinds (expressions used to represent the terms in the theory).  This requires a kinds file ( https://github.com/cvc5/cvc5/blob/main/src/theory/fp/kinds for example ) which gives the names for them, type checking and rewrite rules.  You will also likely need support classes for constants and parser support.  The finite fields theory / PRs by @alex-ozdemir is probably the most recent example of this and will give you an outline of what is needed.

Happy to discuss more if you have some more details of what you want to do.

Cheers,
 - Martin

**Comment by alex-ozdemir** [comment]
Hi @vhavlena,

As Martin alluded to, we're in the process of merging in a theory of (prime-order) finite fields. As we merge it in, I've been drafting a write-up of the steps necessary to create a new theory. I'll publish it after we're done.

In the meantime, you might consult [this wiki article](http://cvc4.cs.stanford.edu/wiki/How_to_write_a_theory_in_CVC4). I followed its instructions to create the finite-field theory solver.

Cheers,
Alex

**Reply by martin-cs** [reply]
I might be pushing my luck here but maybe you could create an example patch-set / branch of how to add a trivial theory?  That way it could be kept up to date with a rebase.

**Reply by alex-ozdemir** [reply]
Yeah, right now I'm planning a blog post that outlines the general steps, with links to the PRs that instantiate those steps for finite fields.

**Reply by daniel-larraz** [reply]
Link to the blog post: https://cvc5.github.io/2024/01/26/adding-a-new-theory-to-cvc5.html

**Comment by vhavlena** [comment]
Hi all,

many thanks for the replies. @martin-cs To be concrete, it is the first case. In particular, we would like to implement own procedure for the theory of strings. In order to get a minimal working example (I don't care about the efficiency right now), is it necessary to provide an implementation of all virtual methods from `theory.h`?

@alex-ozdemir A blog post would be really great! It could help a lot. Also thank for the link, I check it. 

Cheers,
Vojta

**Comment by ajreynol** [comment, marked as answer]
We do not currently have examples of an alternative implementation of an existing theory.

For this purpose, it is worthwhile considering a replacement of the underlying solver for the theory of strings, instead of a new theory solver. Otherwise, the infrastructure for the (current) theory of strings would have to be disabled, since theories cannot share kinds, which would be required assuming you want to use the standard `*.smt2` format for strings.

If you do want to implement an alternative solver for strings (and perhaps one that shares some of the current infrastructure), I could provide some guidance on this.  Longer-term, we could consider having a standard practice for this, which would be significantly less tedious than adding a new theory.
